File: readers-writers.py
'''
NAME: readers-writers.py
AUTHOR: Andrew Meijer V00805554
'''

#!/usr/bin/python3
import threading
import time
import os
import psutil

readers = 0
writing = ""
doneWriting = 0

class writer (threading.Thread):
   def run(self):
      print ("Starting new writer thread.")
      global writing
      w = input("Update the writing: ")

      # ~ critical ~
      roomEmpty.acquire()
      writing = w
      roomEmpty.release()

      print ("Writing is finished.")
      global doneWriting
      doneWriting = 1

class reader (threading.Thread):
    def run(self):
        print ("New reader thread begins.")

        # number of readers in the room is protected
        mutex.acquire()
        global readers
        readers = readers + 1
        if readers == 1:
            roomEmpty.acquire()

        mutex.release()

        # critical section
        print ("There are now ", readers, " readers reading.")
        print('Reader reads "', writing, '" for 2 seconds.')
        time.sleep(2)

        mutex.acquire()
        readers = readers - 1
        if readers == 0:
            roomEmpty.release()
        mutex.release()

# Main
mutex = threading.Lock()
roomEmpty = threading.Lock()

# Create new threads
# Threads can only be started once
r1 = reader()
r2 = reader()
r3 = reader()
w1 = writer()
w2 = writer()

# Start new Threads
# To test readers-writers I will do a few steps:
# Write to shared variable "writing"
# Read with only one reader
# Write again
# Read with 2 readers
w1.start()
r1.start()

# wait
while not doneWriting:
    pass

doneWriting = 0;
w2.start()
r2.start()
r3.start()

# wait until
while not doneWriting:
    pass

print ("Exiting Main Thread.")
# memory usage report:
# nts this is not counted in the comprehensibility measurements
pid = os.getpid()
ps = psutil.Process(pid)
mr = ps.memory_info()
print("Bytes used: ", mr.rss)

File: readers-writers.c
//
// NAME:    readers-writers.c
// AUTHOR:  Andrew Meijer V00805554

//  I use MinGW on Windows10 and compile with the following command:
//      gcc -pthread readers-writers.c

#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

sem_t mutex;
sem_t roomEmpty;
int readers = 0;
//each writer has to tell main it is done.
short doneWriting = 0;
//in this program, the writing is an integer for simplicity.
int writing = 0;

void* writer(void* arg)
{
    int w = 0;
    printf("\nStarting new writer thread.\n");
    printf("Write a number: \n");
    scanf("%d",&w);

    sem_wait(&roomEmpty);
    //critical section
    writing = w;
    sem_post(&roomEmpty);

    printf("Writing is finished.\n");
    doneWriting = 1;
}

void* reader(void* arg)
{
    sem_wait(&mutex);
    readers = readers + 1;
    if(readers == 1){
        sem_wait(&roomEmpty);
    }
    sem_post(&mutex);
    //each reader reads for 2 seconds.
    printf("There are now %d readers reading \"%d\".\n", readers, writing);
    sleep(2);

    sem_wait(&mutex);
    readers = readers - 1;
    if(readers == 0){
        sem_post(&roomEmpty);
    }
    sem_post(&mutex);
}

int main()
{
    sem_init(&mutex, 0, 1);
    sem_init(&roomEmpty, 0, 1);

    pthread_t r1,r2,r3,w1,w2;
    pthread_create(&w1,NULL,writer,NULL);
    pthread_create(&r1,NULL,reader,NULL);

    while(!doneWriting){
        // don't worry. be happy. -_-
    }

    doneWriting = 0;
    pthread_create(&w2,NULL,writer,NULL);
    pthread_create(&r2,NULL,reader,NULL);
    pthread_create(&r3,NULL,reader,NULL);

    while(!doneWriting){}

    sem_destroy(&mutex);
    sem_destroy(&roomEmpty);
    printf("Exiting Main Thread.\n");
    return 0;
}
